core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -318,9 +319,25 @@ private[spark] class SparkSubmit extends Logging {
         args.ivySettingsPath)
 
       if (!StringUtils.isBlank(resolvedMavenCoordinates)) {
-        args.jars = mergeFileLists(args.jars, resolvedMavenCoordinates)
-        if (args.isPython || isInternal(args.primaryResource)) {
-          args.pyFiles = mergeFileLists(args.pyFiles, resolvedMavenCoordinates)
+        // In K8s client mode, when in the driver, add resolved jars early as we might need
+        // them at the submit time. For example we might use the dependencies for downloading
+        // files from a Hadoop Compatible fs eg. S3. In this case the user might pass:
+        // --packages com.amazonaws:aws-java-sdk:1.7.4:org.apache.hadoop:hadoop-aws:2.7.6
+        if (isKubernetesClient &&
+          args.sparkProperties.getOrElse("spark.kubernetes.submitInDriver", "false").toBoolean) {
```

Comment:
what does it mean for `isKubernetesClient` to be true and `spark.kubernetes.submitInDriver` be false?

Reply:
is this like [here](https://github.com/apache/spark/pull/23546/files#r249686501) ?

Reply:
it means its the first submit time in cluster mode. So at this step the submit process needs to upload the dependencies. In practice this is user's machine.